File: java/tests/java8/sodium/MemoryTest3.java
package sodium;

import sodium.*;

public class MemoryTest3
{
    public static void main(String[] args)
    {
        new Thread() {
            public void run()
            {
                try {
                    while (true) {
                        System.out.println("memory "+Runtime.getRuntime().totalMemory());
                        Thread.sleep(5000);
                    }
                }
                catch (InterruptedException e) {
                    System.out.println(e.toString());
                }
            }
        }.start();

        EventSink<Integer> et = new EventSink<Integer>();
        Behavior<Integer> t = et.hold(0);
        EventSink<Integer> eChange = new EventSink<Integer>();
        Behavior<Behavior<Integer>> oout = eChange.map(x -> t).hold(t);
        Behavior<Integer> out = Behavior.switchB(oout);
        Listener l = out.value().listen(tt -> {
            //System.out.println(tt)
        });
        int i = 0;
        while (i < 1000000000) {
            eChange.send(i);
            i++;
        }
        l.unlisten();
    }
}

File: clojure/sodium/src/sodium/memory_test_3.clj
(ns sodium.memory-test-3
  (:require [sodium.core :refer :all])
  (:import [nz.sodium])
  (:import [nz.sodium Operational Stream StreamSink StreamLoop Tuple2 Cell CellSink Transaction Node])
  (:import [java.util Optional])
  (:gen-class))



(defn -main
  "Memory test #3"
  [& args]
  (println "Starting!")
  (.start (Thread. (fn []
                     (loop []
                       (println (str "memory " (.totalMemory (Runtime/getRuntime))))
                       (Thread/sleep 5000)
                       (recur)))))
  (let [et (StreamSink.)
        t (.hold et 0)
        e-change (StreamSink.)
        oout (.hold (.map e-change (apply1 [x] ())) t)
        out (Cell/switchC oout)
        l (.listen out (handler [tt] ()))]
    (loop [i 0]
      (when (< i 1000000000)
        (.send et i)
        (recur (inc i))))
    (.unlisten l))
  (System/exit 0))
  
                     

